src/platform/silabs/tracing/SilabsTracing.cpp

```diff
@@ -0,0 +1,277 @@
+    tracker.mEndTime   = tracker.mStartTime;
+    tracker.mOperation = aOperation;
+    tracker.mType      = OperationType::kInstant;
+    tracker.mError     = error;
+    OutputTrace(tracker);
+}
+
+CHIP_ERROR SilabsTracer::OutputTimeTracker(const TimeTracker & tracker)
+{
+    VerifyOrReturnError(isLogInitialized(), CHIP_ERROR_UNINITIALIZED);
+    ChipLogProgress(DeviceLayer,
+                    "TimeTracker - StartTime: %" PRIu32 ", EndTime: %" PRIu32 ", Operation: %" PRIu8 ", Type: %" PRIu8
+                    ", Error: %" CHIP_ERROR_FORMAT,
+                    tracker.mStartTime.count(), tracker.mEndTime.count(), static_cast<uint8_t>(tracker.mOperation),
+                    static_cast<uint8_t>(tracker.mType), tracker.mError.Format());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SilabsTracer::OutputTrace(const TimeTracker & tracker)
+{
+    // We allow error here as we want to buffer even if the logs are currently uninitialized
+    OutputTimeTracker(tracker);
+
+    if (mBufferedTrackerCount < kMaxBufferedTraces - 1)
+    {
+        mTimeTrackerList.Insert(tracker);
+        mBufferedTrackerCount++;
+    }
+    else
+    {
+        // Save a tracker with 0xFFFF to indicate that the buffer is full
+        TimeTracker resourceExhaustedTracker = tracker;
+        resourceExhaustedTracker.mStartTime  = System::SystemClock().GetMonotonicTimestamp();
+        resourceExhaustedTracker.mEndTime    = resourceExhaustedTracker.mStartTime;
+        resourceExhaustedTracker.mOperation  = TimeTraceOperation::kNumTraces;
+        resourceExhaustedTracker.mType       = OperationType::kInstant;
+        resourceExhaustedTracker.mError      = CHIP_ERROR_BUFFER_TOO_SMALL;
+        mTimeTrackerList.Insert(resourceExhaustedTracker);
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SilabsTracer::OutputWaterMark(TimeTraceOperation aOperation)
+{
+    size_t index     = static_cast<size_t>(aOperation);
+    auto & watermark = mWatermarks[index];
+
+    VerifyOrReturnError(isLogInitialized(), CHIP_ERROR_UNINITIALIZED);
+    ChipLogProgress(DeviceLayer,
+                    "Trace %zu: TotalCount=%" PRIu32 ", SuccessFullCount=%" PRIu32 ", MaxTime=%" PRIu32 ", MinTime=%" PRIu32
+                    ", AvgTime=%" PRIu32 ", CountAboveAvg=%" PRIu32 "",
+                    index, watermark.mTotalCount, watermark.mSuccessfullCount, watermark.mMaxTimeMs.count(),
+                    watermark.mMinTimeMs.count(), watermark.mMovingAverage.count(), watermark.mCountAboveAvg);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SilabsTracer::OutputAllWaterMarks()
+{
+    for (size_t i = 0; i < kNumTraces; ++i)
+    {
+        CHIP_ERROR err = OutputWaterMark(static_cast<TimeTraceOperation>(i));
+        if (err != CHIP_NO_ERROR)
+        {
+            return err;
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SilabsTracer::TraceBufferFlushAll()
```

Comment:
This should be a method of TimeTrackerList.

Reply:
Same as for FlushByOperation.